### PR TITLE
[core] Enable file index for map type with map-keys.

### PR DESCRIPTION
**Pull request by leaves12138**
* API
If you want to create a file index with map-keys, you need to specify.
```sql
CREATE TABLE <PAIMON_TABLE> (<COLUMN> <COLUMN_TYPE> , ...) WITH
(
"file-index.bloom-filter.columns" = "map_data[key1], map_data[key2], map_data[key3], map_data[key4]", --must
"file-index.bloom-filter.map_data.items" = "200", --optional default 1000000
"file-index.bloom-filter.map_data.fpp" = "0.1",  --optional default 0.1
"file-index.bloom-filter.map_data[key1].items" = "10000", --optional
"file-index.in-manifest-threshold" = "500 B" --optional
)
```

If you want to specify the options for specific map key, you can add an external options like:
`file-index.bloom-filter.<map_column>##<map_key>.<op_key>='<op_value>'`
clickhouse: https://clickhouse.com/docs/en/engines/table-engines/mergetree-family/mergetree#table_engine-mergetree-data_skipping-indexes
* NOTE
1. flink sql read `select * from <table> where map_data['key1'] = 'value1'` will not push down
2. Every key in `map-keys` will generate a standalone bloom-filter, so if just need to predicate map_data['key1'] will not all map_data bloom-filter, just only map_data['key1'] bloom-filter.
3. `map-keys` refer to clickhouse "INDEX map_key_index mapKeys(map_column) TYPE bloom_filter GRANULARITY 1"

### CODE REVIEW

**Comment by JingsongLi** [comment]
I prefer `"file-index.bloom-filter.columns" = "map_data[key1]"`.

**Comment by leaves12138** [comment]
> I prefer `"file-index.bloom-filter.columns" = "map_data[key1]"`.

`file-index.bloom-filter.map_data##key1.items` convert to `file-index.bloom-filter.map_data[key1].items` ?

**Comment by leaves12138** [comment]
This pull request is ready for review. @JingsongLi

File: paimon-core/src/main/java/org/apache/paimon/io/FileIndexWriter.java [hunk 1]
```diff
@@ -51,7 +57,8 @@ public final class FileIndexWriter implements Closeable {
     // if the filter size greater than fileIndexInManifestThreshold, we put it in file
     private final long inManifestThreshold;
 
-    private final List<IndexMaintainer> indexMaintainers = new ArrayList<>();
+    private final List<FileIndexMaintainer> fileIndexMaintainers = new ArrayList<>();
+    private final Map<String, MapFileIndexMaintainer> mapFileIndexMaintainers = new HashMap<>();
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/io/FileIndexWriter.java:61]: You should unify fileIndexMaintainers to this Map.

File: paimon-common/src/main/java/org/apache/paimon/options/Options.java [hunk 2]
```diff
@@ -94,6 +102,15 @@ public synchronized String get(String key) {
         return data.get(key);
     }
 
+    public synchronized <X extends Throwable> String getOrThrow(String key, Supplier<X> supplier)
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/options/Options.java:105]: supplier -> exSupplier

File: paimon-common/src/main/java/org/apache/paimon/fileindex/FileIndexOptions.java [hunk 3]
```diff
@@ -44,13 +50,40 @@ public FileIndexOptions(long fileIndexInManifestThreshold) {
     }
 
     public void computeIfAbsent(String column, String indexType) {
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/fileindex/FileIndexOptions.java:52]: You should merge `CoreOptions.indexColumnsOptions` into this class.

And remove `isExternallyPerceptible`.

File: paimon-common/src/main/java/org/apache/paimon/fileindex/FileIndexCommon.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+/** Common function of file index put here. */
+public class FileIndexCommon {
+
+    public static String toMapKey(String mapColumnName, String keyName) {
+        return mapColumnName + "[" + keyName + "]";
+    }
+
+    public static DataType getFieldType(Map<String, DataField> fields, String columnsName) {
+        Optional<Integer> nestedIndex = FileIndexOptions.getNestedColumn(columnsName);
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/fileindex/FileIndexCommon.java:36]: topLevelIndex

Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/fileindex/FileIndexCommon.java:36]: getNestedColumn -> topLevelIndexOfNested

File: paimon-common/src/main/java/org/apache/paimon/fileindex/FileIndexOptions.java [hunk 5]
```diff
@@ -63,7 +96,82 @@ public long fileIndexInManifestThreshold() {
...
-    public Set<Map.Entry<String, Map<String, Options>>> entrySet() {
-        return indexTypeOptions.entrySet();
+    public Set<Map.Entry<Column, Map<String, Options>>> entrySet() {
+        return indexTypeOptions.entrySet().stream()
+                .filter(entry -> entry.getKey().isExternallyPerceptible())
+                .collect(Collectors.toSet());
+    }
+
+    public static Optional<Integer> getNestedColumn(String column) {
+        if (IS_NESTED.matcher(column).find()) {
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/fileindex/FileIndexOptions.java:106]: Can we avoid using `regex.Pattern`?

File: paimon-core/src/main/java/org/apache/paimon/io/FileIndexWriter.java [hunk 6]
```diff
@@ -191,4 +237,76 @@ public byte[] serializedBytes() {
...
+        public String getIndexType() {
+            return indexType;
+        }
+
+        public Map<String, byte[]> serializedBytes() {
+            Map<String, byte[]> result = new HashMap<>();
+            indexWritersMap.forEach(
+                    (k, v) ->
+                            result.put(
+                                    FileIndexCommon.toMapKey(columnName, k), v.serializedBytes()));
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/io/FileIndexWriter.java:308]: You need to consider no value in this file. We should have a special mechanism to compress this situation.

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**